File: lab4/Go/diningPhilosophers.go
package main;

import (
   "fmt"
   "time"
   "sync"
)


type Fork struct {
   mu sync.Mutex
   id int
}


func diningPhilosophers(noOfPhilosophers int, mealsToEat []int) {
   forks := make([]Fork, noOfPhilosophers)
      
   for i := 0; i < noOfPhilosophers; i++ {
      forks[i] = Fork{id : i}
      fmt.Printf("Fork id: %d\n", forks[i].id)
   }
   
   var wg sync.WaitGroup

   for i := 0; i < noOfPhilosophers; i++ {
      wg.Add(1);
      go philosopher(i, &forks[i], &forks[(i + 1) % noOfPhilosophers], mealsToEat[i], &wg)
   }

   wg.Wait();
   time.Sleep(100*time.Millisecond)
}

func philosopher(index int, leftFork *Fork, rightFork *Fork, mealsToEat int, wg *sync.WaitGroup) {
   defer  wg.Done();

   fmt.Printf("Philosopher: %d, (meals left: %d)\n", index, mealsToEat)
   fmt.Printf("Philosopher %d forks: {%d, %d}\n", index, leftFork.id, rightFork.id)

   for {
      time.Sleep(20 * time.Millisecond) //random delay?
      // try to take possession of your left fork
      if leftFork.mu.TryLock() {
         // try to take possession of your right fork
         if rightFork.mu.TryLock() {
            // eat your meal
            fmt.Printf("Philosopher: %d starts eating a meal (meals left: %d)\n", index, mealsToEat);
            mealsToEat = mealsToEat - 1
            // put away right fork
            rightFork.mu.Unlock()
         }
         // put away left fork
         leftFork.mu.Unlock()
         fmt.Printf("Philosopher: %d finished a meal (meals left: %d)\n", index, mealsToEat);
      }

      if mealsToEat == 0 {
         fmt.Printf("Philosopher: %d finished all his meals!\n", index)
         break
      }
   }
}

File: lab4/C/src/diningPhilosophers.c
#include "diningPhilosophers.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h> 


void* eat(void *data)
{
    struct PhilosopherData* philosopher_data = (struct PhilosopherData*) data;
    const int philosopher_idx = philosopher_data->philosopher_idx;
    struct Fork* left_fork = philosopher_data->p_left_fork;
    struct Fork* right_fork = philosopher_data->p_right_fork;
    int* meals_left = &philosopher_data->meals_to_eat;

    while (1)
    {  
        sleep_millis(100);
        printf("Philosopher %d wants to eat      (%d meals to eat left)\n", philosopher_idx, *meals_left);
        
        // try to take left fork
        if (pthread_mutex_trylock(&left_fork->mutex) == 0)
        {
            // try to take right fork
            if (pthread_mutex_trylock(&right_fork->mutex) == 0)
            {
                // eat the meal
                *meals_left -= 1;

        printf("Philosopher %d just ate his meal (%d meals to eat left)\n", philosopher_idx, *meals_left);

                // put down right fork
                pthread_mutex_unlock(&right_fork->mutex);
                // put down left fork
                pthread_mutex_unlock(&left_fork->mutex);
            }
            else
            {
                // put down left fork
                pthread_mutex_unlock(&left_fork->mutex);
            }
        }

        if (*meals_left == 0)
        {
            printf("Philosopher %d finished all of his meals!\n", philosopher_idx);
            break;
        }
    }
    
    return NULL;
}

void join_philosopher_threads(pthread_t *philosophers_array, const int no_of_philosophers)
{
    for (int i = 0; i < no_of_philosophers; ++i)
    {
        pthread_join(philosophers_array[i], NULL);
    }
}

void sleep_millis(const int milliseconds)
{
    const struct timespec ts = {
        milliseconds / 1000,              // seconds
        (milliseconds % 1000) * 1000000   // nanoseconds
    };

    int result = 0;
    do
    {
        result = nanosleep(&ts, NULL);
    } while (result);
}

void clean_up(pthread_t* philosophers, struct Fork* forks, struct PhilosopherData* data)
{
    free(philosophers);
    free(forks);
    free(data);
}

